plugins/modem-manager/fu-mm-device.c

```diff
@@ -931,6 +1116,170 @@ fu_mm_device_write_firmware_mbim_qdu (FuDevice *device, GBytes *fw, GError **err
 }
 
 #endif /* MM_CHECK_VERSION(1,17,1) && MBIM_CHECK_VERSION(1,25,3) */
+
+#if MM_CHECK_VERSION(1,17,2)
+static gboolean
+fu_mm_device_firehose_open (FuMmDevice *self, GError **error)
+{
+	self->firehose_updater = fu_firehose_updater_new (self->port_edl);
+	return fu_firehose_updater_open (self->firehose_updater, error);
+}
+
+static gboolean
+fu_mm_device_firehose_close (FuMmDevice *self, GError **error)
+{
+	g_autoptr(FuFirehoseUpdater) updater = NULL;
+
+	updater = g_steal_pointer (&self->firehose_updater);
+	return fu_firehose_updater_close (updater, error);
+}
+
+static gboolean
+fu_mm_device_firehose_write (FuMmDevice *self, XbSilo *rawprogram_silo,
+			     GPtrArray *rawprogram_actions, GError **error)
+{
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	guint progress_signal_id;
+	gboolean write_result;
+
+	locker = fu_device_locker_new_full (self,
+					    (FuDeviceLockerFunc) fu_mm_device_firehose_open,
+					    (FuDeviceLockerFunc) fu_mm_device_firehose_close,
+					    error);
+	if (locker == NULL)
+		return FALSE;
+
+	progress_signal_id = g_signal_connect_swapped (self->firehose_updater, "write-percentage",
+						       G_CALLBACK (fu_device_set_progress), self);
+
+	write_result = fu_firehose_updater_write (self->firehose_updater,
+						  rawprogram_silo,
+						  rawprogram_actions,
+						  error);
+
+	g_signal_handler_disconnect (self->firehose_updater, progress_signal_id);
+
+	return write_result;
+}
+
+static gboolean
+fu_mm_setup_firmware_dir (FuMmDevice *self, GError **error)
+{
+	g_autofree gchar *cachedir = NULL;
+	g_autofree gchar *mm_fw_dir = NULL;
+
+	/* create a directory to store firmware files for modem-manager plugin */
+	cachedir = fu_common_get_path (FU_PATH_KIND_CACHEDIR_PKG);
+	mm_fw_dir = g_build_filename (cachedir, "modem-manager", "firmware", NULL);
+	if (g_mkdir_with_parents (mm_fw_dir, 0755) == -1) {
```

Comment:
Wondering, `0700` could be better here?

Reply:
probably more secure, yeah. changed to `0700`